src/functional/swing_slide_filter.zig

```diff
@@ -0,0 +1,411 @@
+}
+
+/// Updates the linear function coeficients in `line` that passes throught the two points
+/// of the `segment`. The line is swinged down or up based on the `error_bound`.
+/// If `error_bound` is negative, `line` is swing down. It is swing up otherwise.
+fn updateSwingLine(segment: Segment, line: *LinearFunction, error_bound: f32) void {
+    if (segment.end_time != segment.start_time) {
+        const duration: f80 = @floatFromInt(segment.end_time - segment.start_time);
+        line.slope = (segment.end_value + error_bound - segment.start_value) / duration;
+        line.intercept = segment.start_value - line.slope * usizeToF80(segment.start_time);
+    } else {
+        line.slope = 0.0;
+        line.intercept = segment.start_value;
+    }
+}
+
+/// Evaluate the linear function `line` at `x_time`.
+fn evaluateLineAtTime(line: LinearFunction, x_time: usize) f64 {
+    return @floatCast(line.slope * usizeToF80(x_time) + line.intercept);
+}
+
+/// Append `value` to `compressed_values`.
+fn appendValue(value: f64, compressed_values: *ArrayList(u8)) !void {
+    const valueAsBytes: [8]u8 = @bitCast(value);
+    try compressed_values.appendSlice(valueAsBytes[0..]);
+}
+
+/// Append `index` to `compressed_values`.
+fn appendIndex(index: usize, compressed_values: *ArrayList(u8)) !void {
+    const indexAsFloat: [8]u8 = @bitCast(@as(f64, @floatFromInt(index)));
+    try compressed_values.appendSlice(indexAsFloat[0..]);
+}
+
+/// Computes the numerator of the slope derivate as in Eq. 6.
+fn computeSlopeDerivate(segment: Segment) f80 {
+    return (segment.end_value - segment.start_value) * usizeToF80(segment.end_time - segment.start_time);
+}
+
+/// Compress `uncompressed_values` within `error_bound` using "Swing Filter"
+/// and write the result to `compressed_values`. If an error occurs it is returned.
+pub fn compress_swing(
+    uncompressed_values: []const f64,
+    compressed_values: *ArrayList(u8),
+    error_bound: f32,
+) Error!void {
+    // Create adjusted error bound to avoid reaching the error bound during decompression.
+    const adjusted_error_bound = if (error_bound > 0)
+        error_bound - tersets.ErrorBoundMargin
+    else
+        error_bound;
+
+    var upper_bound_line: LinearFunction = LinearFunction{ .slope = 0, .intercept = 0 };
+    var lower_bound_line: LinearFunction = LinearFunction{ .slope = 0, .intercept = 0 };
+    var new_upper_bound_line: LinearFunction = LinearFunction{ .slope = 0, .intercept = 0 };
+    var new_lower_bound_line: LinearFunction = LinearFunction{ .slope = 0, .intercept = 0 };
+    var current_line: LinearFunction = LinearFunction{ .slope = 0, .intercept = 0 };
+
+    // Initialize the current segment with first two points.
+    var current_segment: Segment = Segment{
+        .start_time = 0,
+        .start_value = uncompressed_values[0],
+        .end_time = 1,
+        .end_value = uncompressed_values[1],
+    };
+
+    // Compute the numerator Eq. (6).
```

Comment:
Either add or remove the parentheses for all cases of "Eq." throughout the file so they are consistent.